moshi/moshi/run_tts.py

```diff
@@ -0,0 +1,210 @@
+
+    def _flush():
+        nonlocal first_batch
+        all_entries = []
+        all_attributes = []
+        prefixes = None
+        if not tts_model.multi_speaker:
+            prefixes = []
+        begin = time.time()
+
+        for request in batch:
+            entries = tts_model.prepare_script(request.turns, padding_between=args.padding_between)
+            all_entries.append(entries)
+            if tts_model.multi_speaker:
+                voices = [tts_model.get_voice_path(voice) for voice in request.voices]
+            else:
+                voices = []
+            all_attributes.append(tts_model.make_condition_attributes(voices, cfg_coef_conditioning))
+            if prefixes is not None:
+                assert len(request.voices) == 1, "For this model, only exactly one voice is supported."
+                prefix_path = hf_get(request.voices[0], args.voice_repo, check_local_file_exists=True)
+                prefixes.append(tts_model.get_prefix(prefix_path))
+
+        print(f"Starting batch of size {len(batch)}")
+        result = tts_model.generate(
+            all_entries, all_attributes, prefixes=prefixes,
+            cfg_is_no_prefix=cfg_is_no_prefix, cfg_is_no_text=cfg_is_no_text)
+        first_batch = False
+        frames = torch.cat(result.frames, dim=-1).cpu()
+        total_duration = frames.shape[0] * frames.shape[-1] / mimi.frame_rate
+        time_taken = time.time() - begin
+        total_speed = total_duration / time_taken
+        print(f"[LM] Batch of size {len(batch)} took {time_taken:.2f}s, "
+              f"total speed {total_speed:.2f}x")
+
+        wav_frames = []
+        with torch.no_grad(), tts_model.mimi.streaming(len(all_entries)):
+            for frame in result.frames[tts_model.delay_steps:]:
+                wav_frames.append(tts_model.mimi.decode(frame[:, 1:]))
+        wavs = torch.cat(wav_frames, dim=-1)
+        effective_duration = 0.
+        for idx, request in enumerate(batch):
+            end_step = result.end_steps[idx]
+            if end_step is None:
+                print(f"Warning: end step is None, generation failed for {request.id}")
+                wav_length = wavs.shape[-1]
+            else:
+                wav_length = int((mimi.sample_rate * (end_step + tts_model.final_padding) / mimi.frame_rate))
+            effective_duration += wav_length / mimi.sample_rate
+            wav = wavs[idx, :, :wav_length]
+            start_step = 0
+            if prefixes is not None:
+                start_step = prefixes[idx].shape[-1]
+                start = int(mimi.sample_rate * start_step / mimi.frame_rate)
+                wav = wav[:, start:]
+            filename = args.out_folder / f"{request.id}.wav"
+            debug_tensors = {
+                'frames': frames[idx].short(),
+            }
+            sphn.write_wav(filename, wav.clamp(-1, 1).cpu().numpy(), mimi.sample_rate)
+            save_file(debug_tensors, filename.with_suffix('.safetensors'))
```

Comment:
Maybe saving the safetensors and json files should be optional and driven by a cli argument?